tasks/task_manager.cc

```diff
@@ -117,48 +179,79 @@ bool task_manager::task::impl::is_done() const noexcept {
     return _status.state == tasks::task_manager::task_state::done;
 }
 
+future<std::vector<task_manager::task::task_essentials>> task_manager::task::impl::get_failed_children() const {
+    return _children.map_each_task<task_essentials>([] (const foreign_task_ptr&) { return std::nullopt; },
+        [] (const task_essentials& child) -> std::optional<task_essentials> {
+            if (child.task_status.state == task_state::failed || !child.failed_children.empty()) {
+                return child;
+            }
+            return std::nullopt;
+        }
+    );
+}
+
 void task_manager::task::impl::run_to_completion() {
-    (void)run().then_wrapped([this] (auto f) {
+    (void)run().then_wrapped([this] (auto f) -> future<> {
         if (f.failed()) {
-            finish_failed(f.get_exception());
+            co_await finish_failed(f.get_exception());
         } else {
+            std::exception_ptr ex = nullptr;
             try {
                 _as.check();
-                finish();
+                co_await finish();
```

Comment:
`finish` is noexcept. I think it's done like this because we can't get the exact exception from `abort_source` without rethrowing it. Please see: https://github.com/scylladb/seastar/pull/2263.

Reply:
You can subscribe a callback that gets a std::exception_ptr parameter and then it will get the error.  Would that help?

Reply:
I don't think so, we don't want to finish task when the callback is triggered, as it will still run

Reply:
ok.  If https://github.com/scylladb/seastar/pull/2263 gets updated in scylla first we can simply this part a bit.

Reply:
otherwise, cleanup can be done as a followup

Reply:
I'm uncomfortable with the complexity here, sorry.
maybe we should simply go with continuations here, like:
```c++
(void)run().then([this] {
    _as.check();
    return finish();
}).handle_exception([this] (std::exception_ptr ex) {
    return finish_failed(std::move(ex));
});
```